exercises/practice/word-count/.approaches/introduction.md

```diff
@@ -0,0 +1,224 @@
+It is similar to writing `Regex(raw"foo")`, but with the benefit that the regex object will be constructed just once (at compile time) rather than on each call to e.g. `wordcount`.
+
+## Approach: Find each word with regex
+
+This is a straightforward algorithm and it is easy to prove that it follows the specification exactly.
+Implementations that split by whitespace and punctuation (see below) tend to count words that the specification does not ask for, e.g. "x1".
+You can view that as a good thing or a bad thing!
+For what it is worth, traditional word counting code like `wc` treat a mix of letters and digits as a word.
+
+```julia
+function wordcount(input)
+    counts = Dict{String, Int}()
+    # Unicode version matching any letters or numbers: r"\pL+'\pL+|\pL+|\pN+"
+    for m in eachmatch(r"[A-Za-z]+'[A-Za-z]+|[A-Za-z]+|[0-9]+", input)
+        word = lowercase(m.match)
+        counts[word] = get(counts, word, 0) + 1
+    end
+    return counts
+end
+```
+
+Same as above, but we lowercase `input` all in one go, so we can have a `Dict` of `SubString`s, which is slightly faster and might be better for some purposes.
+
+```julia
+function wordcount(input)
+    counts = Dict{SubString{typeof(input)}, Int}()
+    # Unicode version matching any letters or numbers: r"\pL+'\pL+|\pL+|\pN+"
+    for m in eachmatch(r"[A-Za-z]+'[A-Za-z]+|[A-Za-z]+|[0-9]+", lowercase(input))
+        word = m.match
+        counts[word] = get(counts, word, 0) + 1
+    end
+    return counts
+end
+```
+
+It's possible to create the dictionary more efficiently with the `countmap` function from `StatsBase`. With `countmap`, you can code-golf the above into:
+
+```julia
+using StatsBase: countmap
+
+wordcount(str) = countmap(m.match for m in eachmatch(r"[A-Za-z]+'[A-Za-z]+|[A-Za-z]+|[0-9]+", lowercase(str)))
+```
+
+`countmap` cheekily uses some implementation details of `Dict` to perform `counts[w] = get(counts, w, 0) + 1` faster than we can (specifically, `countmap` will hash `w` only once).
+
+## Approach: Split, then strip quotation
+
+If we split the input on any character that can't be a word, then we might think that we will only be left with words and words that are surrounded by apostrophes.
+This solution follows that approach.
+
+```julia
+function wordcount(input)
+    words = split(input, r"[^\w']+")
+    counts = Dict{String, Int}()
+    for word in words
+        # Removing leading and closing apostrophes (might be quotation)
+        word = lowercase(strip(word, '''))
+        if !isempty(word)
+            counts[word] = get(counts, word, 0) + 1
+        end
+    end
+    return counts
+end
+```
+
+Note that this solution was written for a previous version of the instructions and counts a lot of words that the new specification does not allow.
+See if you can find some!
+
+<details>
+<summary>Answer</summary>
+
+This solution will count words containing more than one apostrophe as a single word, which is against the specification.
```

Comment:
Do you think this note is also needed in the other solutions that allow for the multiple-apostrophe behaviour? Like the very last one?